src/textual/screen.py

```diff
@@ -404,6 +405,32 @@ def _reset_focus(
         # Go with the what was found.
         self.set_focus(chosen)
 
+    def _update_focus_styles(
+        self, focused: Widget | None = None, blurred: Widget | None = None
+    ) -> None:
+        """Update CSS for focus changes.
+
+        Args:
+            focused: The widget that was focused.
+            blurred: The widget that was blurred.
+        """
+        widgets: set[DOMNode] = set()
+
+        if focused is not None:
+            for widget in reversed(focused.ancestors_with_self):
+                if widget._has_focus_within:
+                    widgets.update(widget.walk_children(with_self=True))
+                    break
+        if blurred is not None:
+            for widget in reversed(blurred.ancestors_with_self):
+                if widget._has_focus_within:
+                    widgets.update(widget.walk_children(with_self=True))
+                    break
+        if widgets:
+            self.app.stylesheet.update_nodes(
+                [widget for widget in widgets if widget._has_focus_within], animate=True
+            )
```

Comment:
Can the first two `if`s evaluate to `True` at the same time?

Also, the third `if` goes over `widgets` and checks `widget._has_focus_within`, which seems to overlap a bit with the inner `if widget._has_focus_within` in each of the first two `if`s.

If this is an optimisation, isn't there a way to short-circuit the conditional in the bottom list comprehension by doing some more work in each of the first two `if`s?
I'm not familiarised with the focus chain so I can't really propose something more specific.

Reply:
> Can the first two `if`s evaluate to `True` at the same time?

Yes, typically when a widget is focused another is blurred.

> 
> Also, the third `if` goes over `widgets` and checks `widget._has_focus_within`, which seems to overlap a bit with the inner `if widget._has_focus_within` in each of the first two `if`s.
> 
> If this is an optimisation, isn't there a way to short-circuit the conditional in the bottom list comprehension by doing some more work in each of the first two `if`s? I'm not familiarised with the focus chain so I can't really propose something more specific.

Focus-within is a weird one.  Focusing a widget can modify the style of an ancestor. Which is why it has to back-track to find a parent with focus-within, then look at the children. It's already a massive win. Don't think there are any significant wins remaning.